[WIP]KAFKA-15444: Native docker image

kagarwal06 opened the pull request:
This PR includes the changes for `native-binary` docker image for Apache Kafka corresponding to the KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-974%3A+Docker+Image+for+GraalVM+based+Native+Kafka+Broker


**NOTE 1** This PR is dependent on https://github.com/apache/kafka/pull/14552, hence should be merged post merging https://github.com/apache/kafka/pull/14552
**NOTE 2:** The changes are made on top of the https://github.com/apache/kafka/pull/14552 as we had common code. Hence this contains changes for the JVM Docker Image as well
**NOTE 3:** To view Native-Image specific changes, refer: https://github.com/VedarthConfluent/kafka/pull/1

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

VedarthConfluent committed:
Add JVM Docker image

VedarthConfluent committed:
Refactor build test and setup proper directory structure

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

kagarwal06 committed:
Merge branch 'docker-image' into native-docker-image

kagarwal06 committed:
KAFKA-15444: Initial Changes for Native Docker Image

VedarthConfluent committed:
Refactor scripts running in Dockerfile

kagarwal06 committed:
KAFKA-15444: Add Licence details

kagarwal06 committed:
Merge branch 'docker-image' into native-docker-image

kagarwal06 committed:
KAFKA-15444: NIT

VedarthConfluent committed:
Refactor bash scripts

kagarwal06 committed:
Merge branch 'docker-image' into native-docker-image

kagarwal06 committed:
KAFKA-15444: Change folder name

VedarthConfluent committed:
Remove redundant files

VedarthConfluent committed:
Add licence to the files

kagarwal06 committed:
Merge branch 'docker-image' into native-docker-image

ijuma commented on core/src/main/scala/kafka/Kafka.scala:
```diff
@@ -86,6 +86,10 @@ object Kafka extends Logging {
+    process(args)
+  }
+
+  def process(args: Array[String]): Unit = {
```
Why is `process` needed? You can invoke `main` from `KafkaNativeWrapper`.

kagarwal06 replied:
Done

ijuma commented on docker/test/requirements.txt:
```diff
@@ -0,0 +1,6 @@
+confluent_kafka
```
This doesn't seem right.

VedarthConfluent replied:
It seems like it's a typo. Updating it to `confluent-kafka`

ijuma replied:
But why would we depend on a confluent kafka artifact.

VedarthConfluent replied:
@ijuma Confluent Kafka for Python - https://github.com/confluentinc/confluent-kafka-python
`confluent-kafka` is a python client compatible with kafka
For sanity tests of docker images we wanted a python client to interact with kafka residing in docker image, hence we went ahead with this publicly available and widely used python client.
We are open to consider alternate python clients if there are any concerns in using this library

ijuma replied:
Can we not use the CLI tools that are included in the Apache Kafka project?

VedarthConfluent replied:
This suggestion has been incorporated. Now we are using CLI scripts included in kafka.

ijuma replied:
Can we remove `confluent_kafka` from here then?

VedarthConfluent replied:
It has been removed. These changes are part of KIP 975 PR raised here:- https://github.com/apache/kafka/pull/14552. (commit 6f64ce54254b985a50b3c5d8d63711cc6b71ab18)
They'll reflect here once this PR is updated

ijuma replied:
Makes sense, thanks.

VedarthConfluent committed:
Updated to 21-jre and fix typos

kagarwal06 committed:
KAFKA-15444: Use main() method directly

VedarthConfluent committed:
Add a release script for pushing docker images to dockerhub

kagarwal06 committed:
Merge branch 'docker-image' into native-docker-image

VedarthConfluent committed:
Update release script to support registry and add error handling

kagarwal06 committed:
Merge branch 'docker-image' into native-docker-image

VedarthConfluent committed:
Remove confluent artifacts from sanity test

Remove confluent kafka python client as dependency
Remove schema registry and connect from compose files
Download kafka binary and use it's scripts to run tests

VedarthConfluent committed:
Remove redundant files

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Fix tests

Add SSL_TOPIC again in constants.py

VedarthConfluent committed:
Use CDS to start kafka

VedarthConfluent committed:
Resolve PR comments

- Remove zookeeper support & tests
- Optimise image by removing redundant packages
- Minor code fixes

VedarthConfluent committed:
Add github actions workflow for build and test of jvm docker image

VedarthConfluent committed:
Add description to kafka url link

VedarthConfluent committed:
Refactors jsa launch script with error handling and timeouts

VedarthConfluent committed:
Remove static sleep from sanity tests

VedarthConfluent committed:
Rely on scripts to detect when server is up

VedarthConfluent committed:
Removed redundant wait in jsa generation

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

kagarwal06 committed:
Merge branch 'docker-image' into native-docker-image

kagarwal06 committed:
KAFKA-15444: Accept kafka-url in the build

VedarthConfluent committed:
Add promotion script

VedarthConfluent committed:
Add requirements.txt for promotion script

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Fix property file location

VedarthConfluent committed:
Add support for supplying properties through file mounting

Changes to add support for file mounting of properties.
Env variables, if defined, will override the file input properties.

VedarthConfluent committed:
Ensure that environment variable configs are always appended in newline

VedarthConfluent committed:
Add missing brackets in template

VedarthConfluent committed:
Add test for file input

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Bubble up test errors to root build test script

VedarthConfluent committed:
Add license comment and refactor ub scripts to remove redundant code

VedarthConfluent committed:
Add readme file and refactor the python scripts

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Add local setup section in Readme

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Update file input test and readme file

VedarthConfluent committed:
Add test for broker metrics

VedarthConfluent committed:
Remove potential flakiness from the test

VedarthConfluent committed:
Remove redundant paths in chown and chmod

VedarthConfluent committed:
Fix build by adding license and excluding files where it cannot be added

VedarthConfluent committed:
Update documentation to include docker image in the release process

VedarthConfluent committed:
Add CDS for storage format

kagarwal06 committed:
Merge branch 'docker-image' into native-docker-image

kagarwal06 committed:
Merge branch 'trunk' into native-docker-image

kagarwal06 committed:
Merge branch 'trunk' into native-docker-image

kagarwal06 committed:
Resolve merge conflicts
